env/posix/ocf_env.c

```diff
@@ -130,8 +131,57 @@ void env_stack_trace(void)
 }
 
 /* *** CRC *** */
-
 uint32_t env_crc32(uint32_t crc, uint8_t const *data, size_t len)
 {
 	return crc32(crc, data, len);
 }
+
+/* *** execution contexts *** */
+pthread_mutex_t exec_context_list_mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t *exec_context_mutex = NULL;
+
+static void init_execution_context(void)
+{
+	unsigned count = env_get_execution_context_count();
+	unsigned i;
+
+	ENV_BUG_ON(count == -1);
+	exec_context_mutex = malloc(count * sizeof(exec_context_mutex[0]));
+	ENV_BUG_ON(exec_context_mutex == NULL);
+	for (i = 0; i < count; i++)
+		ENV_BUG_ON(pthread_mutex_init(exec_context_mutex + i, NULL));
+}
+
+/* get_execuction_context must assure that after the call finishes, the caller
+ * will not get preempted from current execution context. For userspace env
+ * we simulate this behavior by acquiring per execution context mutex. As a
+ * result the caller might actually get preempted, but no other thread will
+ * execute in this context by the time the caller puts current execution ctx. */
+unsigned env_get_execution_context(void)
+{
+	unsigned cpu;
+
+	pthread_mutex_lock(&exec_context_list_mutex);
+	if (exec_context_mutex == NULL)
+		init_execution_context();
+	pthread_mutex_unlock(&exec_context_list_mutex);
+
+	cpu = sched_getcpu();
+	cpu = (cpu == -1) ?  0 : cpu;
+
+	ENV_BUG_ON(pthread_mutex_lock(exec_context_mutex + cpu));
```

Comment:
Using index operator instead of doing explicit pointer arithmetic would look much clearer.